Connectivity drivers

SeppoTakalo opened the pull request:
### Description

Import connectivity drivers into Mbed OS.
* ESP8266 (Wifi)
* Atmel AT86RF233/AT86RF212 (IEEE 802.15.4)
* MCR20A (IEEE 802.15.4)

After this change, any of those driver can be selected as a default driver. This allows building connectivity examples for applications without connectivity.
This allows us to migrate away from easy-connect dependency.

Documentation how default network interfaces work, please see:
* https://github.com/ARMmbed/mbed-os-5-docs/pull/660

This PR depends on https://github.com/ARMmbed/mbed-os/pull/7778

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Breaking change

Commit:
Enable default Mesh interface on KW24D

This device still requires MCR20A driver to be in build tree
and this patch https://github.com/ARMmbed/mcr20a-rf-driver/pull/18

Commit:
Add Atmel RF driver v3.0.4

Commit:
Add MCR20A RF driver v1.0.1

Commit:
Add ESP8266 driver v1.6

SeppoTakalo commented:
@kjbracey-arm please review

jeromecoutant commented:
Hi Seppo

Can I ask why wifi-ism43362 is not imported as well ?

Thx

SeppoTakalo commented:
Hi @jeromecoutant for the first round I'm importing only the ones that we are actively maintaining and testing in our CI.

This is part of the change where I submitted PR earlier into most of the connectivity drivers, ISM as well in https://github.com/ARMmbed/wifi-ism43362/pull/27
Long term goal is to slowly migrate away from easy-connect repository.

Now going forward, applications can be made portable by just choosing which one to provide the default interface.

Driver being outside of Mbed OS tree does not make any difference, once the configuration is in place.

Commit:
Exclude components from Doxygen

Commit:
Remove components from RTOS-less build.

SeppoTakalo commented:
@cmonr What is the justification for "Needs work" label?

0xc0170 commented:
As this is adding components, doesn't this depend on components tools changes or is it good as it is?

SeppoTakalo commented:
@0xc0170 This does not depend on the component support, even that the folder name is the same.

These drivers have been tested not to affect RAM or ROM usage, as long as they are unused. So there is no need to exclude from the build.

SeppoTakalo commented:
Rebased on top of master.
MCR20A driver changed so that it is enabled if build against KW24D target.

cmonr commented:
@SeppoTakalo I would normally say to check the failure logs, but 1) the job artifacts have been garbage collected by Jenkins due to the history length, and 2) I'm not sure how to retrigger a build.

@ARMmbed/mbed-os-test Some help?

AnttiKauppila commented:
This is dependent on https://github.com/ARMmbed/mbed-os/pull/7778 which should go in first

SeppoTakalo commented:
Client test is known to fail, because it is testing with previous release.
Once this goes in, there are fixes to client in the internal repositories. PRs already submitted.

cmonr commented:
@SeppoTakalo @0xc0170 Keep in mind that @bulislaw is on vacation for the rest of the month.

> @cmonr What is the justification for "Needs work" label?

I think it was because the cloud_client_smoke_test check was failing. Still talking with @ARMmbed/mbed-os-test to see how to restart the test.

cmonr commented:
Now waiting on @kjbracey-arm and the predeeding PR.

SeppoTakalo commented:
Rebased on top of master.

SeppoTakalo commented:
I accidentally rebased on top of my working copy, so this contains fixes from #7778 as well.

cmonr commented:
@SeppoTakalo On the plus side, if #7778 happens to be ready and passes CI, this PR can immediately start CI w/o a rebase being needed.

AnttiKauppila commented:
@0xc0170 Can you build this?

0xc0170 commented:
The build queue looks fine, testing not so will take some time

/morph build

donatieng commented on components/802.15.4_RF/atmel-rf-driver/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*~
```
That `.gitignore` file should be removed

donatieng commented on components/802.15.4_RF/mcr20a-rf-driver/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*~
```
Same here

SeppoTakalo replied:
These are verbatim copy of the existing repositories, but maybe you are right. We should clean up the stuff that we don't need.

donatieng reviewed (approved):
Looks good to me but there are a few `.gitignore` files that should probably be removed

0xc0170 commented:
Examples causing multipledefinition - do they need an update? Is this dependency that needs to be resolved ?

0xc0170 commented:
> Examples causing multipledefinition - do they need an update? Is this dependency that needs to be resolved ?

Please rebase as well, one file conflicts now

0xc0170 commented:
> This PR depends on #7778

This will go after, so postponing this one

cmonr commented:
After a rebase, we'll put in CI posthaste!

0xc0170 commented:
^^ This needs attention asap, will chase this

Commit:
On KW24D and NCS36510 default to smaller Nanostack configuration

These boards do not fit the full Nanostack, so we need to select
6LoWPAN-ND as a default.

0xc0170 commented:
Can you review travis docs failure (was not there before)

Commit:
Remove unneeded .gitignore files

SeppoTakalo commented:
Fixed the doxyfile problem. Was a conflict with another commit.

@0xc0170 This should now be ready for testing.
Example application (wifi, socket & mesh) modified to support this.

0xc0170 commented:
I was just about to comment , as was running doxygen locally and found the bug in doxygen file but not fixed, thanks !

/morph build

0xc0170 commented:
Various failures related to the changeset. Please review

Commit:
TEMPORARY Skip building of examples that rely on drivers.

Commit:
Fix AT24MAC driver builds on targets that device SDA/SCL names

Some targets have SDA and SCL defined in public headers, breaking the
build of AT24MAC driver.

SeppoTakalo commented:
AT24MAC driver should be now fixed on some of the platforms.. I'll continue to study if there was other failures.

Commit:
WISE-1510 is not Arduino form factor

Commit:
Add dummy delay loop for ARMC6

This needs proper implementation later.

cmonr commented:
/morph build

Commit:
SDT64B is not Arduino FF. It does not define Arduino pins.

Commit:
RF drivers need DEVICE_SPI

cmonr commented:
/morph build

SeppoTakalo commented:
Added one more protective rule.. part of MCR20A driver was not flagged by `DEVICE_SPI` so I added that as well.

SeppoTakalo commented:
@cmonr Please restart the tests job.

SeppoTakalo commented:
/morph build

cmonr commented:
@SeppoTakalo This is gona need a rebase in addition to code/target fixes.

Commit:
Disable mesh networking on tests by default

Testing a mesh requires proper environment.

SeppoTakalo commented:
Rebased on top of master to fix the conflict.
Needs rebuild now

/morph build

cmonr commented:
/morph test

SeppoTakalo commented:
@cmonr what is the test problem? There seems to be no log file.

0xc0170 commented:
`time_start`  KeyError, reported already to test team this week. The test was aborted to be restarted

SeppoTakalo commented:
@0xc0170 Will you restart the test?

0xc0170 commented:
@SeppoTakalo Can you rebase please? and start the build right away

cmonr commented:
@SeppoTakalo Looks like I was able to do the rebase.

/morph build

cmonr commented:
I derp'd and used origin/master instead of upstream/master. Doing the rebase again.

cmonr commented:
/morph build

pan- commented on components/802.15.4_RF/atmel-rf-driver/source/NanostackRfPhyAtmel.cpp:
```diff
@@ -321,6 +321,13 @@ __asm static void delay_loop(uint32_t count)
   BX   lr
 }
+#elif defined (__ARMCC_VERSION) /* ARMC6 */
+void delay_loop(uint32_t count)
```
In which world this doesn't get optimized out ([ref](https://godbolt.org/z/nl3tfn)) ?

SeppoTakalo replied:
It does, I just needed a placeholder to get it building with ARMC6.

Other compilers already have assembly version provided, but none of those builded OK with ARMC6

geky commented:
Should this be in a COMPONENT_ESP8266 directory?
